pkg/controllers/jobset_controller_test.go

```diff
@@ -1174,6 +1182,1042 @@ func jobWithFailedCondition(name string, failureTime time.Time) *batchv1.Job {
+					jobName:           "test-jobset-replicated-job-1-test-job",
+					ns:                ns,
+					replicas:          1,
+					jobIdx:            0}).
+					Parallelism(1).DeletionTimestamp(&metav1.Time{Time: time.Now()}).
+					Completions(2).Suspend(true).
+					Ready(1).
+					Active(1).Obj(),
+			},
+			expectedJobsDeleted: nil,
+		},
+		{
+			name: "delete jobs with error",
+			jobs: []*batchv1.Job{
+				makeJob(&makeJobArgs{
+					jobSetName:        jobSetName,
+					replicatedJobName: "replicated-job-1",
+					jobName:           "test-jobset-replicated-job-1-test-job",
+					ns:                ns,
+					replicas:          1,
+					jobIdx:            0}).
+					Parallelism(1).
+					Completions(2).Suspend(true).
+					Ready(1).
+					Active(1).Obj(),
+			},
+			expectedJobsDeleted: nil,
+			errorFlag:           true,
+			expectedErr:         errors.Join(errors.New("delete pod error")),
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			var jobsDeleted []*batchv1.Job
+			fc := makeClient(interceptor.Funcs{
+				Delete: func(ctx context.Context, client client.WithWatch,
+					obj client.Object, opts ...client.DeleteOption) error {
+					job, ok := obj.(*batchv1.Job)
+					if !ok {
+						return nil
+					}
+					if tc.errorFlag || job == nil {
+						return errors.New("delete pod error")
+					}
+					job = tc.jobs[0]
+					jobsDeleted = append(jobsDeleted, job)
+					return nil
+				},
+			})
+			r := &JobSetReconciler{
+				Client: fc.client,
+				Scheme: fc.scheme,
+				Record: fc.record,
+			}
+			actual := r.deleteJobs(context.TODO(), tc.jobs)
+			assert.Equal(t, actual, tc.expectedErr)
+
+			sort.Slice(jobsDeleted, func(i, j int) bool {
+				return jobsDeleted[i].Name < jobsDeleted[j].Name
+			})
+			sort.Slice(tc.expectedJobsDeleted, func(i, j int) bool {
+				return tc.expectedJobsDeleted[i].Name < tc.expectedJobsDeleted[j].Name
+			})
+
+			if !reflect.DeepEqual(jobsDeleted, tc.expectedJobsDeleted) {
+				t.Errorf("deleteJobs() did not make the expected job deletion calls")
+			}
+		})
+	}
+}
+
+func TestJobSetFinished(t *testing.T) {
+	var (
+		jobSetName = "test-jobset"
+		ns         = "default"
+	)
+	tests := []struct {
+		name     string
+		js       *jobset.JobSet
+		expected bool
+	}{
+		{
+			name: "test jobSetFinished with not finished",
+			js: testutils.MakeJobSet(jobSetName, ns).
+				ReplicatedJob(testutils.MakeReplicatedJob("replicated-job-1").
+					Job(testutils.MakeJobTemplate("test-job", ns).Obj()).
+					Replicas(1).
+					Obj()).Obj(),
+		},
+		{
+			name: "test jobSetFinished with finished",
+			js: testutils.MakeJobSet(jobSetName, ns).
+				SetConditions([]metav1.Condition{{Type: string(jobset.JobSetCompleted), Status: metav1.ConditionTrue}}).
+				ReplicatedJob(testutils.MakeReplicatedJob("replicated-job-1").
+					Job(testutils.MakeJobTemplate("test-job", ns).Obj()).
+					Replicas(1).
+					Obj()).Obj(),
+			expected: true,
+		},
+	}
+	for _, tc := range tests {
```

Comment:
Can you add a test case for a failed JobSet? It should be considered finished as well.